pkg/resource/generator.go

```diff
@@ -99,13 +100,54 @@ func (g *Generator) syncStorage(cr *imageregistryv1.Config) error {
 	}
 
 	if runCreate {
+		reconf := g.storageReconfigured(cr)
 		if err := driver.CreateStorage(cr); err != nil {
 			return err
 		}
+		if reconf {
+			metrics.StorageBackendReconfigured()
+		}
 	}
 	return nil
 }
 
+// storageReconfigured returns true if we are, based on the provided config,
+// starting to use a new underlying storage location. This function compares
+// the previous storage location with the new one returning true if we have a
+// previous configuration and it differs from the current one. Previous storage
+// configuration is extracted from Status while current comes from Spec. If
+// storage is not configured false is returned.
+func (g *Generator) storageReconfigured(cr *imageregistryv1.Config) bool {
+	prev := cr.Status.Storage
+	cur := cr.Spec.Storage
+
+	// If we cannot find storage configuration on our Status it means we
+	// are creating the storage for the first time. On this case we simply
+	// return false.
+	prevStorageID, err := storage.ID(prev)
```

Comment:
I'm concerned about the switches inside `ID` and `Engine`, someone may forget to update it.

We can use newDriver here and add `ID() string` to our `Driver` interface. If types are different (`reflect.TypeOf(prevDriver) != reflect.TypeOf(curDriver)`) then we know for sure that the storage is reconfigured. If types are the same, we can use `prevDriver.ID() != curDriver.ID()`.

@ricardomaraschini what do you think?

Reply:
That would work if we first solve #394 Chicken and egg now.

Reply:
It is implemented. PTAL.